[export] inline jit.scripted function in export

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155381
* __->__ #155180

When we export a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

We do the same thing for script method, where a "\_\_wrapped\_\_" attribute points to the original callable in most cases. There are some corner cases we identified: top-level jit.scripted modules' method doesn't have a \_\_wrapped\_\_. In this case, we fall back to the original scripted approach. Maybe there're more such cases but need verification.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

ydwu4 committed:
[export] inline jit.scripted function in export

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1485,12 +1485,29 @@ void initJitScriptBindings(PyObject* module) {
+                    .cast<bool>()) {
+              TORCH_CHECK(
+                  py::hasattr(args[0], py::str("_torchdynamo_inline")),
+                  "When export scripted function ",
```
```suggestion
                  "During PT2 Exporting, we encountered TorchScripted function. To trace through it, 
```

tugsbayasgalan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1485,12 +1485,29 @@ void initJitScriptBindings(PyObject* module) {
+              // remove the function itself with args[1:]
+              py::slice slice0(1, args.size(), 1);
+              return args[0].attr("_torchdynamo_inline")(
+                  *args[slice0], **kwargs);
```
Does **kwargs work in C++?

ydwu4 replied:
yeah, it works.

tugsbayasgalan reviewed (commented):
cc: @justinchuby

tugsbayasgalan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1485,12 +1485,29 @@ void initJitScriptBindings(PyObject* module) {
-            return result;
+            if (py::module::import("torch")
+                    .attr("compiler")
+                    .attr("is_exporting")()
```
Does this cover non-strict export?

ydwu4 commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1485,12 +1485,29 @@ void initJitScriptBindings(PyObject* module) {
+                  "Please file an issue to PyTorch if you see this error.");
+
+              // remove the function itself with args[1:]
+              py::slice slice0(1, args.size(), 1);
```
tuple_slice gives a pytorch internal data structure, we expect a py::tuple here so i just use the plain slicing.

ydwu4 commented on test/export/test_export.py:
```diff
@@ -616,6 +616,24 @@ def forward(self, x, y):
 
+    def test_inline_script_function(self):
+        @torch.jit.script
+        def _forward(x: torch.Tensor):
```
sure!

ydwu4 replied:
see test_inline_script_class_method and test_inline_script_method

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we exporting a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

justinchuby commented:
> > Does this handle control flows of the scripted function
> 
> It handles the control flow the same way as we exporting the python code: we specialize the constants and shapes, errors out for data-dependent ones.

I think it can be more helpful: the scripted functions are usually data dependent with control flows

ydwu4 commented:
Yeah, I agree. Maybe we could try to use the auto converter tool we're developing when we see  a jit.script annotation is added. It's also a good test bed for testing the coverage of the auto converter.

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we exporting a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we export a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

We do the same thing for script method, where a "\_\_wrapped\_\_" attribute points to the original callable in most cases. There are some corner cases we identified: top-level jit.scripted modules' method doesn't have a \_\_wrapped\_\_. In this case, we fall back to the original scripted approach. Maybe there're more such cases but need verification.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we export a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

We do the same thing for script method, where a "\_\_wrapped\_\_" attribute points to the original callable in most cases. There are some corner cases we identified: top-level jit.scripted modules' method doesn't have a \_\_wrapped\_\_. In this case, we fall back to the original scripted approach. Maybe there're more such cases but need verification.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we export a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

We do the same thing for script method, where a "\_\_wrapped\_\_" attribute points to the original callable in most cases. There are some corner cases we identified: top-level jit.scripted modules' method doesn't have a \_\_wrapped\_\_. In this case, we fall back to the original scripted approach. Maybe there're more such cases but need verification.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] inline jit.scripted function in export"


When we export a scripted function, we inline the original callable stored in "_torchdynamo_inline", this is the same strategy as torch.compile path. 

We do the same thing for script method, where a "\_\_wrapped\_\_" attribute points to the original callable in most cases. There are some corner cases we identified: top-level jit.scripted modules' method doesn't have a \_\_wrapped\_\_. In this case, we fall back to the original scripted approach. Maybe there're more such cases but need verification.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]
